File: cmd/module.go
// Package cmd implements the Cobra CLI commands for elmos.
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

// moduleCmd - kernel module management
var moduleCmd = &cobra.Command{
	Use:   "module",
	Short: "Manage kernel modules",
	Long: `Build, load, and manage out-of-tree kernel modules.

Modules are stored in the modules/ directory and can be built
against the configured kernel, then loaded in QEMU via 9p share.`,
}

var moduleBuildCmd = &cobra.Command{
	Use:   "build [name]",
	Short: "Build kernel modules",
	Long:  `Build one or all kernel modules. If no name specified, builds all.`,
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := ctx.EnsureMounted(); err != nil {
			return err
		}
		name := ""
		if len(args) > 0 {
			name = args[0]
		}
		return runModuleBuild(name)
	},
}

var moduleCleanCmd = &cobra.Command{
	Use:   "clean [name]",
	Short: "Clean module build artifacts",
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := ctx.EnsureMounted(); err != nil {
			return err
		}
		name := ""
		if len(args) > 0 {
			name = args[0]
		}
		return runModuleClean(name)
	},
}

var moduleStatusCmd = &cobra.Command{
	Use:   "status",
	Short: "Show module build status",
	RunE: func(cmd *cobra.Command, args []string) error {
		return runModuleStatus()
	},
}

var moduleListCmd = &cobra.Command{
	Use:   "list",
	Short: "List available modules",
	RunE: func(cmd *cobra.Command, args []string) error {
		return runModuleList()
	},
}

var moduleNewCmd = &cobra.Command{
	Use:   "new [name]",
	Short: "Create a new module from template",
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		return runModuleNew(args[0])
	},
}

var moduleHeadersCmd = &cobra.Command{
	Use:   "headers",
	Short: "Prepare kernel headers for module building",
	RunE: func(cmd *cobra.Command, args []string) error {
		if err := ctx.EnsureMounted(); err != nil {
			return err
		}
		return runModuleHeaders()
	},
}

func init() {
	moduleCmd.AddCommand(moduleBuildCmd)
	moduleCmd.AddCommand(moduleCleanCmd)
	moduleCmd.AddCommand(moduleStatusCmd)
	moduleCmd.AddCommand(moduleListCmd)
	moduleCmd.AddCommand(moduleNewCmd)
	moduleCmd.AddCommand(moduleHeadersCmd)
}

func runModuleBuild(name string) error {
	cfg := ctx.Config

	// Get list of modules to build
	modules, err := getModules(name)
	if err != nil {
		return err
	}

	if len(modules) == 0 {
		printInfo("No modules found to build")
		return nil
	}

	for _, modName := range modules {
		modPath := filepath.Join(cfg.Paths.ModulesDir, modName)

		printStep("Building module: %s", modName)

		cmd := exec.Command("make",
			"-C", cfg.Paths.KernelDir,
			fmt.Sprintf("M=%s", modPath),
			fmt.Sprintf("ARCH=%s", cfg.Build.Arch),
			"LLVM=1",
			fmt.Sprintf("CROSS_COMPILE=%s", cfg.Build.CrossCompile),
			"modules",
		)
		cmd.Env = ctx.GetMakeEnv()
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr

		if err := cmd.Run(); err != nil {
			printError("Failed to build module: %s", modName)
			return err
		}

		printSuccess("Built: %s", modName)
	}

	return nil
}

func runModuleClean(name string) error {
	cfg := ctx.Config

	modules, err := getModules(name)
	if err != nil {
		return err
	}

	for _, modName := range modules {
		modPath := filepath.Join(cfg.Paths.ModulesDir, modName)

		printStep("Cleaning module: %s", modName)

		cmd := exec.Command("make",
			"-C", cfg.Paths.KernelDir,
			fmt.Sprintf("M=%s", modPath),
			fmt.Sprintf("ARCH=%s", cfg.Build.Arch),
			"clean",
		)
		cmd.Stdout = os.Stdout
		cmd.Stderr = os.Stderr

		if err := cmd.Run(); err != nil {
			printWarn("Failed to clean module: %s", modName)
		}
	}

	printSuccess("Modules cleaned")
	return nil
}

func runModuleStatus() error {
	cfg := ctx.Config

	modules, _ := getModules("")
	if len(modules) == 0 {
		printInfo("No modules found")
		return nil
	}

	fmt.Println()
	fmt.Printf("  %-20s %-12s\n", "MODULE", "STATUS")
	fmt.Println("  " + strings.Repeat("-", 35))

	for _, modName := range modules {
		modPath := filepath.Join(cfg.Paths.ModulesDir, modName)
		koFile := filepath.Join(modPath, modName+".ko")

		status := "not built"
		if _, err := os.Stat(koFile); err == nil {
			status = successStyle.Render("✓ built")
		}

		fmt.Printf("  %-20s %s\n", modName, status)
	}

	fmt.Println()
	return nil
}

func runModuleList() error {
	cfg := ctx.Config

	modules, _ := getModules("")
	if len(modules) == 0 {
		printInfo("No modules found in %s", cfg.Paths.ModulesDir)
		return nil
	}

	fmt.Println("Available modules:")
	for i, mod := range modules {
		modPath := filepath.Join(cfg.Paths.ModulesDir, mod)
		srcFile := filepath.Join(modPath, mod+".c")

		desc := ""
		if content, err := os.ReadFile(srcFile); err == nil {
			// Extract MODULE_DESCRIPTION
			lines := strings.Split(string(content), "\n")
			for _, line := range lines {
				if strings.Contains(line, "MODULE_DESCRIPTION") {
					start := strings.Index(line, "\"")
					end := strings.LastIndex(line, "\"")
					if start >= 0 && end > start {
						desc = line[start+1 : end]
					}
					break
				}
			}
		}

		fmt.Printf("  %d. %s", i+1, mod)
		if desc != "" {
			fmt.Printf(" - %s", desc)
		}
		fmt.Println()
	}

	return nil
}

func runModuleNew(name string) error {
	cfg := ctx.Config

	modPath := filepath.Join(cfg.Paths.ModulesDir, name)

	// Check if already exists
	if _, err := os.Stat(modPath); err == nil {
		return fmt.Errorf("module already exists: %s", name)
	}

	// Create directory
	if err := os.MkdirAll(modPath, 0755); err != nil {
		return err
	}

	// Create source file
	srcContent := fmt.Sprintf(`// SPDX-License-Identifier: GPL-2.0
/*
 * %s - Kernel module
 */

#include <linux/init.h>
#include <linux/module.h>
#include <linux/kernel.h>

static int __init %s_init(void)
{
    pr_info("%s: Module loaded\n");
    return 0;
}

static void __exit %s_exit(void)
{
    pr_info("%s: Module unloaded\n");
}

module_init(%s_init);
module_exit(%s_exit);

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Your Name");
MODULE_DESCRIPTION("A simple kernel module");
MODULE_VERSION("1.0");
`, name, name, name, name, name, name, name)

	srcPath := filepath.Join(modPath, name+".c")
	if err := os.WriteFile(srcPath, []byte(srcContent), 0644); err != nil {
		return err
	}

	// Create Makefile
	makeContent := fmt.Sprintf(`obj-m += %s.o

# Optional: Add extra source files
# %s-objs := %s.o helper.o
`, name, name, name)

	makePath := filepath.Join(modPath, "Makefile")
	if err := os.WriteFile(makePath, []byte(makeContent), 0644); err != nil {
		return err
	}

	printSuccess("Created module: %s", modPath)
	printInfo("Edit %s/%s.c to implement your module", modPath, name)
	return nil
}

func runModuleHeaders() error {
	printStep("Preparing kernel headers for module building...")

	// Run modules_prepare
	jobs := ctx.Config.Build.Jobs
	return runBuild(jobs, []string{"modules_prepare"})
}

func getModules(name string) ([]string, error) {
	cfg := ctx.Config

	if name != "" {
		// Check specific module exists
		modPath := filepath.Join(cfg.Paths.ModulesDir, name)
		if _, err := os.Stat(modPath); os.IsNotExist(err) {
			return nil, fmt.Errorf("module not found: %s", name)
		}
		return []string{name}, nil
	}

	// Get all modules
	entries, err := os.ReadDir(cfg.Paths.ModulesDir)
	if err != nil {
		return nil, err
	}

	var modules []string
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		name := entry.Name()
		// Check for Makefile
		makePath := filepath.Join(cfg.Paths.ModulesDir, name, "Makefile")
		if _, err := os.Stat(makePath); err == nil {
			modules = append(modules, name)
		}
	}

	return modules, nil
}

File: src/domain/builder/module.cpp
// ============================================================================
// domain/builder/module.cpp — Kernel module build orchestration
// ============================================================================

#include "module.hpp"

#include <config/types.hpp>
#include <context/context.hpp>
#include <domain/toolchain/manager.hpp>

#include "kernel.hpp"

#include <filesystem>
#include <sstream>

namespace elmos::domain::builder {

namespace fs = std::filesystem;

ModuleBuilder::ModuleBuilder(context::Context* ctx, toolchain::Manager* tm) : ctx_(ctx), tm_(tm) {}

auto ModuleBuilder::build(std::stop_token token, const std::string& name) -> VoidResult {
    auto modules = get_modules(name);
    if (!modules)
        return make_error(modules.error());
    for (const auto& mod : *modules) {
        auto r = build_module(token, mod);
        if (!r)
            return r;
    }
    return {};
}

auto ModuleBuilder::build_module(std::stop_token token, const ModuleInfo& mod) -> VoidResult {
    auto& cfg = ctx_->config();
    auto env = ctx_->get_make_env();

    return ctx_->exec().run_with_env(token, env, "make",
                                     {
                                         "-C",
                                         cfg.paths.kernel_dir,
                                         "M=" + mod.path,
                                         "ARCH=" + cfg.build.arch,
                                         "LLVM=1",
                                         "modules",
                                     });
}

auto ModuleBuilder::clean(std::stop_token token, const std::string& name) -> VoidResult {
    auto modules = get_modules(name);
    if (!modules)
        return make_error(modules.error());
    auto& cfg = ctx_->config();
    for (const auto& mod : *modules) {
        ctx_->exec().run(token, "make",
                         {
                             "-C",
                             cfg.paths.kernel_dir,
                             "M=" + mod.path,
                             "ARCH=" + cfg.build.arch,
                             "clean",
                         });
    }
    return {};
}

auto ModuleBuilder::get_modules(const std::string& name) -> Result<std::vector<ModuleInfo>> {
    auto& cfg = ctx_->config();
    if (!ctx_->fs().exists(cfg.paths.modules_dir)) {
        return std::vector<ModuleInfo>{};
    }

    if (!name.empty()) {
        auto path = (fs::path(cfg.paths.modules_dir) / name).string();
        if (!ctx_->fs().exists(path)) {
            return make_error(Error::generic("module not found: " + name));
        }
        return std::vector{get_module_info(name, path)};
    }

    auto entries = ctx_->fs().read_dir(cfg.paths.modules_dir);
    if (!entries)
        return make_error(entries.error());

    std::vector<ModuleInfo> modules;
    for (const auto& entry : *entries) {
        if (!entry.is_directory)
            continue;
        auto path = (fs::path(cfg.paths.modules_dir) / entry.name).string();
        auto make_path = (fs::path(path) / "Makefile").string();
        if (!ctx_->fs().exists(make_path))
            continue;
        modules.push_back(get_module_info(entry.name, path));
    }
    return modules;
}

auto ModuleBuilder::get_module_info(const std::string& name, const std::string& path)
    -> ModuleInfo {
    ModuleInfo info{.name = name, .path = path};
    auto ko = (fs::path(path) / (name + ".ko")).string();
    info.built = ctx_->fs().exists(ko);

    auto src = (fs::path(path) / (name + ".c")).string();
    auto content = ctx_->fs().read_file(src);
    if (content) {
        info.description = extract_description(*content);
    }
    return info;
}

auto ModuleBuilder::extract_description(const std::string& content) -> std::string {
    std::istringstream stream(content);
    std::string line;
    while (std::getline(stream, line)) {
        if (line.find("MODULE_DESCRIPTION") != std::string::npos) {
            auto start = line.find('"');
            auto end = line.rfind('"');
            if (start != std::string::npos && end > start) {
                return line.substr(start + 1, end - start - 1);
            }
        }
    }
    return "";
}

auto ModuleBuilder::prepare_headers(std::stop_token token) -> VoidResult {
    auto& cfg = ctx_->config();
    auto env = ctx_->get_make_env();
    return ctx_->exec().run_with_env(token, env, "make",
                                     {
                                         "-C",
                                         cfg.paths.kernel_dir,
                                         "-j" + std::to_string(cfg.build.jobs),
                                         "ARCH=" + cfg.build.arch,
                                         "LLVM=1",
                                         "modules_prepare",
                                     });
}

auto ModuleBuilder::create_module(const std::string& name) -> VoidResult {
    auto& cfg = ctx_->config();
    auto mod_path = (fs::path(cfg.paths.modules_dir) / name).string();

    if (ctx_->fs().exists(mod_path)) {
        return make_error(Error::generic("module already exists: " + name));
    }

    auto r = ctx_->fs().mkdir_all(mod_path);
    if (!r)
        return r;

    // Create minimal module source
    std::string c_name = name;
    for (auto& c : c_name) {
        if (c == '-')
            c = '_';
    }

    std::string src = "#include <linux/init.h>\n#include <linux/module.h>\n\n"
                      "static int __init " +
                      c_name +
                      "_init(void) {\n"
                      "    pr_info(\"" +
                      name +
                      " loaded\\n\");\n    return 0;\n}\n\n"
                      "static void __exit " +
                      c_name +
                      "_exit(void) {\n"
                      "    pr_info(\"" +
                      name +
                      " unloaded\\n\");\n}\n\n"
                      "module_init(" +
                      c_name +
                      "_init);\n"
                      "module_exit(" +
                      c_name +
                      "_exit);\n\n"
                      "MODULE_LICENSE(\"GPL\");\n"
                      "MODULE_DESCRIPTION(\"A simple kernel module\");\n";

    auto wr = ctx_->fs().write_file((fs::path(mod_path) / (c_name + ".c")).string(), src);
    if (!wr)
        return wr;

    std::string makefile = "obj-m := " + c_name + ".o\n";
    return ctx_->fs().write_file((fs::path(mod_path) / "Makefile").string(), makefile);
}

}  // namespace elmos::domain::builder
